pkg/cfn/builder/api.go

```diff
@@ -0,0 +1,197 @@
+package builder
+
+import (
+	"encoding/base64"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	cfn "github.com/aws/aws-sdk-go/service/cloudformation"
+	gfn "github.com/awslabs/goformation/cloudformation"
+
+	"github.com/kubicorn/kubicorn/pkg/logger"
+)
+
+const (
+	awsStackName  = "AWS::StackName"
+	awsAccountID  = "AWS::AccountId"
+	awsRegion     = "AWS::Region"
+	fnSub         = "Fn::Sub"
+	fnJoin        = "Fn::Join"
+	fnSplit       = "Fn::Split"
+	fnGetAtt      = "Fn::GetAtt"
+	fnImportValue = "Fn::ImportValue"
+
+	ParamClusterName      = "ClusterName"
+	ParamClusterStackName = cfnOutputClusterStackName
+	ParamNodeGroupID      = "NodeGroupID"
+
+	clusterTemplateDescription                = "EKS cluster"
+	clusterTemplateDescriptionDefaultFeatures = " (with dedicated VPC & IAM role) "
+
+	nodeGroupTemplateDescription                = "EKS nodes"
+	nodeGroupTemplateDescriptionDefaultFeatures = " (Amazon Linux 2 with SSH) "
+
+	templateDescriptionSuffix = " [created and managed by eksctl]"
+)
+
+type resourceSet struct {
+	template *gfn.Template
+	outputs  []string
+}
+
+func newResourceSet() *resourceSet {
+	return &resourceSet{
+		template: gfn.NewTemplate(),
+	}
+}
+
+func makeRef(refName string) *gfn.StringIntrinsic {
+	return gfn.NewStringRef(refName)
+}
+
+func makeSub(expr string) *gfn.StringIntrinsic {
+	return gfn.NewStringIntrinsic(fnSub, expr)
+}
+
+func makeName(suffix string) *gfn.StringIntrinsic {
+	return makeSub(fmt.Sprintf("${%s}-%s", awsStackName, suffix))
+}
+
+func makeSlice(i ...*gfn.StringIntrinsic) []*gfn.StringIntrinsic {
+	return i
+}
+
+func makeStringSlice(s ...string) []*gfn.StringIntrinsic {
+	slice := []*gfn.StringIntrinsic{}
+	for _, i := range s {
+		slice = append(slice, gfn.NewString(i))
+	}
+	return slice
+}
+
+func makeAutoNameTag(suffix string) gfn.Tag {
+	return gfn.Tag{
+		Key:   gfn.NewString("Name"),
+		Value: makeSub(fmt.Sprintf("${%s}/%s", awsStackName, suffix)),
+	}
+}
+
+var refStackName = makeRef(awsStackName)
+
+func (r *resourceSet) newParameter(name, valueType, defaultValue string) *gfn.StringIntrinsic {
+	p := map[string]string{"Type": valueType}
+	if defaultValue != "" {
+		p["Default"] = defaultValue
+	}
+	r.template.Parameters[name] = p
+	return makeRef(name)
+}
+
+func (r *resourceSet) newStringParameter(name, defaultValue string) *gfn.StringIntrinsic {
+	return r.newParameter(name, "String", defaultValue)
+}
+
+func (r *resourceSet) newResource(name string, resource interface{}) *gfn.StringIntrinsic {
+	r.template.Resources[name] = resource
+	e := reflect.ValueOf(resource).Elem()
```

Comment:
Needs a comment – it for setting `Name` tag on all resources that have tags. It also appends resource type to the value (which is what CloudFormation cannot do for us, and it seems like the kind of things we everyone does and wants to see).